docs/cpp_docs.rst

```diff
@@ -26,94 +44,556 @@ The beginning of both header and source code files should include
+
+Images are allowed, but should not be central to the explanation; users viewing the documentation as text must be able to comprehend its meaning without resorting to an image viewer.
+These additional illustrations are included using:
+
+.. code-block:: cpp
+
+   /**
+    * @image html filename ["caption"]
+    */
+
+where ``filename`` is a path relative to the project root directory.
+
+.. _cpp-doxygen-references:
+
+References
+----------
+
+References can be included either in the :ref:`'Notes' <cpp-doxygen-notes>` section or in a separate list below them. A reference consists of the ``@cite`` tag, followed by a BibTeX label. Bibfiles must be listed in the ``CITE_BIB_FILES`` configuration tag in ``doc/doxygen.conf.in``.
+
+Note that Web pages should be referenced with regular inline links.
+
+References are meant to augment the documentation, but should not be required to understand it.
+
+.. _cpp-doxygen-examples:
+
+Examples
+--------
+
+'Examples' is an optional section for examples. This section is very strongly encouraged.
+
+Examples should be prefixed by ``@code`` and suffixed by ``@endcode``:
+
+.. code-block:: cpp
+
+   /**
+    * @code
+    * auto cosines = cos(angles);
+    * @endcode
+    *
+    * Comment explaining the second example.
+    *
+    * @code
+    * auto cosines = cos(radians(angles));
+    * @endcode
+    */
+
+File Description Comment for Header Files
+=========================================
+
+``.h`` files should not contain documentation on the file itself (i.e., a documentation block containing a ``@file`` command). Document the class or package instead, where possible.
+
+.. _cpp-doxygen-package-definition:
+
+Documenting/Defining Packages
+==============================
 
 Each LSST package corresponds to a group in Doxygen.
 We declare this package in the root header file for a package, usually named ``package.h``.
 
-In this header file, below the preamble, provide a Doxygen comment block that declares the package with the fields:
+In this header file, below the boilerplate but above any ``#include`` statements or other code, provide a Doxygen comment block that declares the package with the fields:
 
-1. ``@defgroup`` followed by the ``PackageName`` and  ``PackageTitle``
+1. ``@defgroup`` followed by machine-readable and human-readable names for the package
+2. :ref:`cpp-doxygen-short-summary` (must be prefixed by ``@brief`` because the summary is not the first line.)
+3. :ref:`cpp-doxygen-deprecation` (if applicable)
+4. :ref:`cpp-doxygen-extended-summary` (recommended)
+5. :ref:`cpp-doxygen-see-also` (optional)
 
-2. ``@brief`` to provide a one-line description of the package.
+The extended summary should include the `preferred abbreviation <https://developer.lsst.io/coding/cpp_style_guide.html#names-representing-namespaces-must-be-camelcase-with-leading-lowercase-letter-and-based-on-component-or-directory-name>`_ for the package's namespace, if one exists.
```

Comment:
Maybe a question for @jonathansick: how does this package summary relate to the Python summary?  It seems like they'd contain essentially the same content, and should perhaps fill in one from the other?